### DISCUSSION TITLE
Easy way to see RGB values sent to each LED?

### DISCUSSION THREAD

**Discussion by chcore** [awawa-dev/HyperHDR#1000] [Q&A]
Pretty much what the title says. I'd like to see the RGB values being sent to the LEDs in real time. Is that possible? Maybe it could be a nice addition to the live preview window?

**Comment by awawa-dev** [comment]
LED Hardware -> File and then use some text editor to view the output file. I use Far Manager to view it in the real time.

**Reply by chcore** [reply]
Thanks, that works.

I notice that if I am displaying a single color uniformly on my screen such as a purple rgb(150,75,225) I cannot seem to get that same value (or something nearing it) to be pushed to the LEDs.

- With gamma set to 1.5 (default), the LEDs receive rgb(105,24,198)
- With gamma set to 1.0, the LEDs receive rgb(141,55,216)
- A screenshot grabbed from the live preview has the color as rgb(143,57,218) which is rather close to the 1.0 gamma value

So I suppose the next question is, does the live preview contain the exact frames as received from the grabber, or is it after applying LUT or other transformations? Should the LEDs not be receiving something close to the original rgb(150,75,225) value? For what it's worth, they're all still fairly purple.

I will note that for now I am only concerned with SDR content, the source is NV12 from an LG TV webos grabber and my calibration settings look like this:

```
  "adjustment": {
    "white": [ 255, 255, 255 ],
    "red": [ 255, 0, 0 ],
    "green": [ 0, 255, 0 ],
    "blue": [ 0, 0, 255 ],
    "cyan": [ 0, 255, 255 ],
    "magenta": [ 255, 0, 255 ],
    "yellow": [ 255, 255, 0 ],
    "backlightThreshold": 0,
    "backlightColored": false,
    "brightness": 100,
    "brightnessCompensation": 0,
    "gammaRed": 1.0,
    "gammaGreen": 1.0,
    "gammaBlue": 1.0
  }
```

**Reply by awawa-dev** [reply]
> So I suppose the next question is, does the live preview contain the exact frames as received from the grabber, or is it after applying LUT or other transformations? Should the LEDs not be receiving something close to the original rgb(150,75,225) value? For what it's worth, they're all still fairly purple.

The video frame remains "the same" in a sense, but encoded in mjpeg format with a very high degree of compression and is sent to the browser in this form.

> I will note that for now I am only concerned with SDR content, the source is NV12 from an LG TV webos grabber and my calibration settings look like this:

Did you encode this color into yuv420 limited SDR video test file? During calibration you can enable "Debug" option and after calibration check in logs the closest to selected color test point to know what to expect later. BTW webOS captured images are targeting much darker (than standard USB grabbers) display brightness.

**Reply by chcore** [reply]
> Did you encode this color into yuv420 limited SDR video test file?

I didn't, but there looks to be plenty of similar purple colors already in there.

> During calibration you can enable "Debug" option and after calibration check in logs the closest to selected color test point to know what to expect later

I'll try that, thanks. I will also try out LCH and see if there is an improvement.

**Reply by chcore** [reply]
> The video frame remains "the same" in a sense, but encoded in mjpeg format with a very high degree of compression and is sent to the browser in this form.

The current LUT does affect what is displayed in the live preview frame. With the default generated LUT, I get rgb(176,0,255) whereas with a calibrated LUT I get rgb(149,50,237). Changing color channel settings (e.g. gamma) has no such effect though.

**Reply by awawa-dev** [reply]
Yes, it's the same frame as it's processed in HyperHDR. So if the tone mapping is enabled it affects both LEDs and live preview.

**Reply by chcore** [reply]
OK so I decided to test using a color that already exists in the SDR test video. The LUT translates this color (and everything else!) correctly, according to the log:

`[144  80 224] => [144  80 225]`

When I fill the entire screen with rgb(144,80,225) I can see the LEDs are actually receiving rgb(143,56,232) with my gamma values set to 1.0, or rgb(107,26,221) when set to 1.5.

Is this normal behavior? Shouldn't the LEDs be receiving rgb(144,80,225) instead?

**Reply by awawa-dev** [reply]
The calibration process targeting MP4 files&player&TV, that color is present in the MP4 test files because it's divided by 16 so can have a approximation in the result. If you inject the color in any other way or mixing calibration 420 limited MP4 with color injected in 420 full range MP4 then there is no guarantee how webOS or any grabber will process that color later: must be tested to confirm it.

**Reply by chcore** [reply]
> The calibration process targeting MP4 files&player&TV, that color is present in the MP4 test files because it's divided by 16 so can have a approximation in the result. If you inject the color in any other way or mixing calibration 420 limited MP4 with color injected in 420 full range MP4 then there is no guarantee how webOS or any grabber will process that color later.

Interesting. My method of filling the screen with rgb(144,80,225) is generating a 4k solid PNG image using ImageMagick, and viewing that image in Kodi. Does that still apply?

**Reply by awawa-dev** [reply]
use ffmpeg and create test file with solids boards just like in the latest source directory

**Comment by chcore** [comment]
I used the following command and the attached rgb(144,80,224) image to create the attached test video. For me, the color is off in the test video, no matter where I play it (kodi, mpv, chrome, etc). Is that expected?

`ffmpeg -loop 1 -t 30 -framerate 1/3 -i rgb-144-80-224.png -shortest -map 0:v:0 -vf fps=25,colorspace=space=bt709:primaries=bt709:range=pc:trc=iec61966-2-1:ispace=bt709:iprimaries=bt709:irange=pc:itrc=iec61966-2-1:format=yuv444p12:fast=0:dither=none -c:v libx265 -pix_fmt yuv420p10le -profile:v main10 -preset veryslow -x265-params keyint=75:min-keyint=75:bframes=0:scenecut=0:psy-rd=0:psy-rdoq=0:rdoq=0:sao=false:cutree=false:deblock=false:strong-intra-smoothing=0:lossless=1:colorprim=bt709:transfer=iec61966-2-1:colormatrix=bt709:range=full -f mp4 rgb-144-80-224.mp4`

https://github.com/user-attachments/assets/280bf2be-b0be-40f5-983f-d098aa06fdcd

![rgb-144-80-224](https://github.com/user-attachments/assets/4dcd9f46-9e7a-4ed4-988b-08184c478073)

**Reply by awawa-dev** [reply]
I cannot open this video in my players

**Reply by chcore** [reply]
Wonder why that is, as I used a command very similar to the one in your run_me.bat (basically just excluded the audio stream).

What if you generate the MP4 on your machine?

**Reply by awawa-dev** [reply]
maybe github transcoded it? did you verify upload?

**Reply by chcore** [reply]
Yes, just downloaded it and the checksum matches the original.

**Reply by awawa-dev** [reply]
Probably thats the problem (paletted PNG):
![obraz](https://github.com/user-attachments/assets/2833cdc1-acf6-4bc6-bf49-2518e20515f8)

**Reply by chcore** [reply]
I converted the PNG into a BMP first, then provided that as input for ffmpeg. Getting the same result.

**Reply by awawa-dev** [reply]
BMP can be palleted as well. Use Gimp and export precised 24-bit PNG.

**Reply by chcore** [reply]
Same result with an 8bpc RGB PNG (assuming that is the correct one). Here is the source image and output.

https://github.com/user-attachments/assets/a62a23f3-d4f2-4adf-a118-914b7c8c3896

![rgb-144-80-224](https://github.com/user-attachments/assets/913fc0a3-775f-4c82-9daf-4a7d5ee9adf4)

**Reply by awawa-dev** [reply]
And what color do you read?

**Reply by chcore** [reply]
> And what color do you read?

The LEDs receive rgb(139,83,210) at gamma 1.0.

Actually reasonably close, but not as good as what the log would indicate? Ideally it would be rgb(144,80,225).

**Reply by awawa-dev** [reply]
Hmm...I think I probably know what's happening but to confirm it could you prepare a testboard for color white (255,255,255)?

**Reply by chcore** [reply]
LEDs receive rgb(236,236,236) with this rgb(255,255,255) test video:

https://github.com/user-attachments/assets/6aed613d-3454-4216-ab76-bad1e5ad455f

**Reply by awawa-dev** [reply]
Yep. The player threats YUV (16-235) limited range file as a full range YUV (0-255). Probably you also could not achieve the black color (0,0,0) but (16,16,16).

**Reply by awawa-dev** [reply]
On the other hand webOS recognizes it correctly but I want to see calibration logs.

**Reply by chcore** [reply]
> Yep. The player threats YUV (16-235) limited range file as a full range YUV (0-255). Probably you also could not achieve the black color (0,0,0) but (16,16,16).

That is correct, I just created a 0,0,0 test video and the LEDs receive 16,16,16. It also looks grey on-screen rather than pure black.

I'll run the calibration again and provide a log.

**Reply by chcore** [reply]
> On the other hand webOS recognizes it correctly but I want to see calibration logs.

Attached: [hyperhdr-log.txt](https://github.com/user-attachments/files/18000529/hyperhdr-log.txt)

**Reply by awawa-dev** [reply]
Thank you. So it's still YUV limted on webOS. And HyperHDR adjusts the range on its own. With score around 10 you have almost 1:1 mapping for SDR. But I dont know why players threads SDR YUV limited as a full range? Maybe some new tag is needed?

**Reply by chcore** [reply]
> Yep. The player threats YUV (16-235) limited range file as a full range YUV (0-255). Probably you also could not achieve the black color (0,0,0) but (16,16,16).

When I play the files in chrome and MPV on my desktop (no hyperhdr installed), they are also wrong there. A color picker shows chrome playing the pure white file at 236,236,236 while MPV has it at 218,218,218.

**Reply by awawa-dev** [reply]
You can always use 2 full range test files for SDR.

**Reply by chcore** [reply]
I won't pretend to understand any of it, but maybe this: https://trac.ffmpeg.org/wiki/colorspace

**Reply by chcore** [reply]
> You can always use 2 full range test files for SDR.

Two?

**Reply by awawa-dev** [reply]
Yes, I used it to prepare these files. Maybe limited range yuv is not a best target or something is wrong in the transformation but I dont know what can it be. And what about full range SDRs?

**Reply by awawa-dev** [reply]
> > You can always use 2 full range test files for SDR.
> 
> Two?

https://github.com/awawa-dev/awawa-dev.github.io/tree/master/calibration/misc_test_files

[test_SDR_yuv420_low_quality.mp4](https://github.com/awawa-dev/awawa-dev.github.io/blob/master/calibration/misc_test_files/test_SDR_yuv420_low_quality.mp4)
[test_SDR_yuv444_high_quality.mp4](https://github.com/awawa-dev/awawa-dev.github.io/blob/master/calibration/misc_test_files/test_SDR_yuv444_high_quality.mp4)

**Reply by chcore** [reply]
> > > You can always use 2 full range test files for SDR.
> > 
> > 
> > Two?
> 
> https://github.com/awawa-dev/awawa-dev.github.io/tree/master/calibration/misc_test_files
> 
> [test_SDR_yuv420_low_quality.mp4](https://github.com/awawa-dev/awawa-dev.github.io/blob/master/calibration/misc_test_files/test_SDR_yuv420_low_quality.mp4) [test_SDR_yuv444_high_quality.mp4](https://github.com/awawa-dev/awawa-dev.github.io/blob/master/calibration/misc_test_files/test_SDR_yuv444_high_quality.mp4)

Are the black parts of these test videos supposed to be pure black when played? They don't look pure black when played in Kodi.

**Reply by chcore** [reply]
> [test_SDR_yuv420_low_quality.mp4](https://github.com/awawa-dev/awawa-dev.github.io/blob/master/calibration/misc_test_files/test_SDR_yuv420_low_quality.mp4)
> [test_SDR_yuv444_high_quality.mp4](https://github.com/awawa-dev/awawa-dev.github.io/blob/master/calibration/misc_test_files/test_SDR_yuv444_high_quality.mp4)

I was already using these (rather, the yuv420 one) until you updated the repo and moved them into the misc folder.

**Reply by chcore** [reply]
> [test_SDR_yuv420_low_quality.mp4](https://github.com/awawa-dev/awawa-dev.github.io/blob/master/calibration/misc_test_files/test_SDR_yuv420_low_quality.mp4)
> [test_SDR_yuv444_high_quality.mp4](https://github.com/awawa-dev/awawa-dev.github.io/blob/master/calibration/misc_test_files/test_SDR_yuv444_high_quality.mp4)

Calibrated using both of these, neither correctly sent 144,80,224 to the LEDs. The yuv444 LUT sent 142,36,243 while yuv420 sent 143,56,232.

**Reply by awawa-dev** [reply]
OK, give me more time. I will investigate the SDR issue. Probably colorspace plugin cannot perform color range conversion or the container is missing the tag. HDR processing is using different plugin.

**Comment by awawa-dev** [comment]
Hi @chcore 
I managed to fixed the SDR files by using limited/full range tag for colorspace. For zscale it is unnecessary.
Your color is of course present in the test files from 0:12 to 0:15, first top row as on the snapshot.
I used VLC and took the screenshot (Video->Take Snapshot) 
![obraz](https://github.com/user-attachments/assets/2d43b5b3-346f-49fb-a13e-feac62725250)
So the (144,80,224) is (144,79,224) in the calibration file. Thats also explains why we never will achieve 0 score in the calibration process ;-) but I think we are OK with that. There is a small noise inside this rectangle both for YUV420 limited and full range, YUV444 has almost no noise.

**Reply by chcore** [reply]
Thanks @awawa-dev

The test_SDR_yuv444_high_quality.mp4 file still looks like it's limited range to me in Kodi (no pure blacks). I will try out test_SDR_yuv420_low_quality.mp4 (which looks like full range edit: scratch that, it looks the same as the previous commit, which was already less washed out than 444) in the mean time and see how it goes.

**Reply by chcore** [reply]
After calibrating with test_SDR_yuv444_high_quality.mp4 I get a great result with the videos generated in https://github.com/awawa-dev/HyperHDR/discussions/1000#discussioncomment-11454865 and https://github.com/awawa-dev/HyperHDR/discussions/1000#discussioncomment-11454427 - 141,78,221. That is very close to the expected color 144,80,225.

But when viewing 144,80,225 in the form of a PNG image (attached to the same comments above), the LEDs receive 146,36,243.

So it looks like the calibration only works correctly for limited range content?

**Reply by awawa-dev** [reply]
Try to encode the PNG image into MP4 using updated commands in the source dir.

**Reply by awawa-dev** [reply]
What is the value of WHITE_0 now? Previously it was set incorrect too high (29 instead of 16)  by incorrect SDR test file but fixed later in the calibration:
>     WHITE_0: [  0   0   0] => [ 15  15  15] (YUV: [ 29 128 128])

**Reply by chcore** [reply]
Here's a log of calibration with test_SDR_yuv444_high_quality.mp4: 
[hyperhdr-log.txt](https://github.com/user-attachments/files/18009617/hyperhdr-log.txt)

**Reply by awawa-dev** [reply]
I suspect it's a player issue now, mpc-hc with madvr just like VLC also plays the file correctly now: the screenshot was taken using windows printscreen but you read black & white colors in the last row: (0,0,0) and (255,255,255)
![obraz](https://github.com/user-attachments/assets/ef1a647e-e32f-4cf7-879e-419802bd7b59)

**Reply by awawa-dev** [reply]
Seems in Kodi it's a bit complicated https://kodi.wiki/view/Video_levels_and_color_space#Limited_vs_Full_Video_Levels
HDR forces some standard at least: only limited range is usually supported by a hardware acceleration

**Reply by chcore** [reply]
What do you think is the solution?

The reason I am testing with RGB images in addition to video is I would eventually like to be able to play video games with the same setup (which I think are not going to be limited range).

**Reply by awawa-dev** [reply]
I think it's a Kodi range configuration problem, not only for the calibration file but for all yuv420 limited files

> WHITE_0: [  0   0   0] => [ 29  29  29] (YUV: [ 41 128 128]), [ 28  28  28] (YUV: [ 40 128 128])  [source noise detected]
it should be YUV [16 128 128]

One of user currently testing this case on webOS for me, will let you know about the result.

**Reply by chcore** [reply]
> One of user currently testing this case on webOS for me, will let you know about the result.

Thank you. In case it is relevant, I have the TV, GPU and Kodi all set to full range.

**Reply by awawa-dev** [reply]
So according to Kodi documentation it's wrong. I'm not sure, I dont use it. Content (test file) is limited so it should be rather limited => full => limited
![obraz](https://github.com/user-attachments/assets/b9e58a08-5d2e-4940-9a8e-c51dd7cd59fd)

**Reply by chcore** [reply]
I also noticed that, but it says Full,Full,Full will have desktop (i.e. video games) and video both looking correct, which is what I'm after.

I might give limited,full,limited a go but it seems like that's the best option for a device that will only ever be playing video.

**Reply by chcore** [reply]
I played around with limited-full-limited and various test videos but couldn't get anything that ends up sending the right color to the LEDs for the purple PNG image (i.e. non-video content).

**Comment by awawa-dev** [comment]
@chcore
1 After testing we decided to revert changes introduced in https://github.com/awawa-dev/HyperHDR/discussions/1000#discussioncomment-11461118 . Colorspace worked just fine for webOS.
2 For testing RGB444 12bit mode was used (SDR) windows and VLC in fullscreen mode.
3 Calibration shows correct WHITE_0 range: 16 no 29 or higher.
```
2024-12-04T23:46:30.459Z [CALIBRATOR] (LutCalibrator.cpp:381) Using flatbuffers/protobuffers as a sourceWaiting for first captured test board..
2024-12-04T23:46:30.479Z [CALIBRATOR] The YUV range is LIMITED
Captured colors:

     WHITE_0: [  0   0   0] => [  0   0   0] (YUV: [ 16 128 128]) 
     WHITE_1: [ 16  16  16] => [ 14  14  14] (YUV: [ 28 128 128]) 
     WHITE_2: [ 32  32  32] => [ 28  28  28] (YUV: [ 40 128 128]) 
     WHITE_3: [ 48  48  48] => [ 43  43  43] (YUV: [ 53 128 128]) 
     WHITE_4: [ 64  64  64] => [ 57  57  57] (YUV: [ 65 128 128]) 
     WHITE_5: [ 80  80  80] => [ 72  72  72] (YUV: [ 78 128 128]) 
     WHITE_6: [ 96  96  96] => [ 86  86  86] (YUV: [ 90 128 128]) 
     WHITE_7: [112 112 112] => [101 101 101] (YUV: [103 128 128]) 
     WHITE_8: [128 128 128] => [116 116 116] (YUV: [116 128 128]) 
     WHITE_9: [144 144 144] => [132 132 132] (YUV: [129 128 128]) 
    WHITE_10: [160 160 160] => [146 146 146] (YUV: [141 128 128]) 
    WHITE_11: [176 176 176] => [161 161 161] (YUV: [154 128 128]) 
    WHITE_12: [192 192 192] => [176 176 176] (YUV: [167 128 128]) 
    WHITE_13: [208 208 208] => [191 191 191] (YUV: [180 128 128]) 
    WHITE_14: [224 224 224] => [205 205 205] (YUV: [192 128 128]) 
    WHITE_15: [240 240 240] => [220 220 220] (YUV: [205 128 128]) 
    WHITE_16: [255 255 255] => [234 234 234] (YUV: [217 128 128]) 
```

and purple is calibrated as below:
`[144  80 224] => [144  82 227] `

4 After calibration and forwarded LEDs output to the file that purple test file was played:
[purple_limited_range_SDR_yuv420.zip](https://github.com/user-attachments/files/18015123/purple_limited_range_SDR_yuv420.zip)

5 The purple result in the output file:

![obraz](https://github.com/user-attachments/assets/3204b8a9-c647-4773-8d8d-2eae93f2eced)

6 From my point of view it closes the subject, but I'm open for further ffmpeg command improvements.
Probably it's mismatch range configuration in Kodi but I dont use it so cant support it.
BTW make sure you have tone mapping enabled for SDR content on webOS when testing: after calibration it can be turned off for flatbuffers when USB grabber was not excluded from build (PC/Rpi versions).

**Reply by chcore** [reply]
@awawa-dev thank you so much for looking into it. I will review kodi/driver/OS configurations and see if I can produce the right results.

**Reply by chcore** [reply]
I thought I would try running the calibration just by opening the PNG images (the ones used to make the HyperHDR test videos) in Kodi and manually skipping through them so that HyperHDR can collect the data. To my surprise, it all worked. The colors for SDR are now very accurate.

I'll see if I can produce similar results for HDR.

**Reply by awawa-dev** [reply]
You would achieve a similar effect much more easily by selecting the second calibration mode via the browser in SDR mode. You are still missing one important thing in your Kodi setup: by displaying RGB images you are completely ignoring what happens with YUV material described here: https://kodi.wiki/view/Video_levels_and_color_space#Limited_vs_Full_Video_Levels.

**Reply by chcore** [reply]
> You would achieve a similar effect much more easily by selecting the second calibration mode via the browser in SDR mode. You are still missing one important thing in your Kodi setup: by displaying RGB images you are completely ignoring what happens with YUV material described here: https://kodi.wiki/view/Video_levels_and_color_space#Limited_vs_Full_Video_Levels.

Agreed, will play with it.
